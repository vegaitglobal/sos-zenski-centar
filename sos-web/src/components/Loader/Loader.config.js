/**
 * Configuration for the spinner
 */
export const SIZE = 45;
export const STROKE_WIDTH = 4;
export const RADIUS = SIZE / 2 - STROKE_WIDTH * 2;
export const OUTLINE_SIZE = Math.PI * (2 * RADIUS);
