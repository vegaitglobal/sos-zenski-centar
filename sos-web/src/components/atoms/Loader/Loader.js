import {
  StyledAnimatingCircle,
  StyledBaseCircle,
  StyledLoader,
  StyledLoaderHolder,
} from './Loader.styles';

export const Loader = (props) => (
  <StyledLoaderHolder {...props}>
    <StyledLoader>
      <StyledBaseCircle />
      <StyledAnimatingCircle />
    </StyledLoader>
  </StyledLoaderHolder>
);
