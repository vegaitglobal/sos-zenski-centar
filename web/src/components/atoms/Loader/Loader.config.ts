/**
 * Configuration for the spinner
 */
export const SIZE = 60;
export const STROKE_WIDTH = 2;
export const RADIUS = SIZE / 2 - STROKE_WIDTH * 2;
export const OUTLINE_SIZE = Math.PI * (2 * RADIUS);
