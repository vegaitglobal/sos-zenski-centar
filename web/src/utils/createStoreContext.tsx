import { createContext, PropsWithChildren, FunctionComponent, Context, useContext } from 'react';
import useStoreFactory from '../hooks/useStoreFactory';

/**
 * Creates a context for the store created by the passed `storeFactory`.
 * Additionally returns the `Provider` component and the hook for the created store's context.
 *
 * @param name - name of the store
 * @param storeFactory - a function that returns an object representing the store
 */
// eslint-disable-next-line @typescript-eslint/naming-convention
const createStoreContext = <TStore extends Record<string, unknown>>(
	name: string,
	storeFactory: () => TStore,
): [FunctionComponent<unknown>, () => TStore, Context<TStore>] => {
	const context = createContext<TStore>({} as TStore); // the initial value is only to satisfy TS and won't be used

	/**
	 * Initializes and contains the context for the `TStore`.
	 */
	function StoreProvider({ children }: PropsWithChildren<unknown>): JSX.Element {
		const store = useStoreFactory(storeFactory);
		return <context.Provider value={store}>{children}</context.Provider>;
	}

	const useStore = (): TStore => {
		const store = useContext(context);
		if (store) {
			return store;
		}
		throw new ReferenceError(
			`${name} is not initialized, make sure that your component is wrapped in the 'Provider' of the store.`,
		);
	};

	return [StoreProvider, useStore, context];
};

export default createStoreContext;
