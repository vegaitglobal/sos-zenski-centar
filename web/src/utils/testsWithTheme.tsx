import { render, RenderOptions, RenderResult } from '@testing-library/react';
import { ThemeProvider } from 'styled-components';
import { ReactElement, ComponentType, ReactNode } from 'react';
import { theme } from '../styles/config/theme';

export const componentWithTheme = (component: ReactNode): ReactElement => (
	<ThemeProvider theme={theme}>{component}</ThemeProvider>
);

export const renderWithTheme = (component: ReactElement, options?: RenderOptions): RenderResult => {
	const wrapper: ComponentType = ({ children }) => componentWithTheme(children);

	return render(component, { wrapper, ...options });
};
