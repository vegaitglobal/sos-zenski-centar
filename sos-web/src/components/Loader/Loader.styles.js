import styled, { keyframes } from 'styled-components';
import { OUTLINE_SIZE, RADIUS, SIZE, STROKE_WIDTH } from './Loader.config';

const rotate = keyframes`
  100% {
    transform: rotate(360deg);
  }
`;

const dash = keyframes`
  0% {
    stroke-dasharray: 0, ${OUTLINE_SIZE};
    stroke-dashoffset: 0;
  }
  50% {
    stroke-dasharray: ${OUTLINE_SIZE / 2} ${OUTLINE_SIZE / 2};
    stroke-dashoffset: 0;
  }
  100% {
    stroke-dasharray: 0, ${OUTLINE_SIZE};
    stroke-dashoffset: -${OUTLINE_SIZE};
  }
`;

export const StyledLoaderHolder = styled.div`
  width: 100%;
  text-align: center;
  padding: 1rem 0;
`;

export const StyledLoader = styled.svg.attrs(() => ({
  width: SIZE,
  height: SIZE,
}))`
  animation: ${rotate} 2s linear infinite;
`;

const StyledCircle = styled.circle.attrs(() => ({
  cx: SIZE / 2,
  cy: SIZE / 2,
  r: RADIUS,
}))`
  stroke-width: ${STROKE_WIDTH}px;
  fill: none;
`;

export const StyledBaseCircle = styled(StyledCircle)`
  stroke: ${({ theme }) => theme.color.pink};
`;

export const StyledAnimatingCircle = styled(StyledCircle)`
  stroke-dasharray: 50, ${OUTLINE_SIZE};
  stroke-dashoffset: 0;
  animation: ${dash} 1.5s ease-in-out infinite;
  stroke: ${({ theme }) => theme.color.purple};
`;
