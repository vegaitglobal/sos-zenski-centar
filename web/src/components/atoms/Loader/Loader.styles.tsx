import styled from 'styled-components';
import { rotate, dash } from '../../../styles/config/keyframes';

import { OUTLINE_SIZE, RADIUS, SIZE, STROKE_WIDTH } from './Loader.config';

export const StyledLoaderHolder = styled.div`
	width: 100%;
	text-align: center;
	padding: 1rem 0;
`;

export const StyledLoader = styled.svg.attrs(() => ({
	width: SIZE,
	height: SIZE,
}))`
	animation: ${rotate} 2s linear infinite;
`;

const StyledCircle = styled.circle.attrs(() => ({
	cx: SIZE / 2,
	cy: SIZE / 2,
	r: RADIUS,
}))`
	stroke-width: ${STROKE_WIDTH}px;
	fill: none;
`;

export const StyledBaseCircle = styled(StyledCircle)`
	stroke: ${({ theme }) => theme.color.whiteFaded};
`;

export const StyledAnimatingCircle = styled(StyledCircle)`
	stroke-dasharray: 50, ${OUTLINE_SIZE};
	stroke-dashoffset: 0;
	animation: ${dash} 1.5s ease-in-out infinite;
	stroke: ${({ theme }) => theme.color.black};
`;
