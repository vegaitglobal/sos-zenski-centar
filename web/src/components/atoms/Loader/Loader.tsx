import { ReactElement } from 'react';
import { LoaderProps } from './Loader.data';
import {
	StyledAnimatingCircle,
	StyledBaseCircle,
	StyledLoader,
	StyledLoaderHolder,
} from './Loader.styles';

export const Loader = (props: LoaderProps): ReactElement => (
	<StyledLoaderHolder data-testid="Loader" {...props}>
		<StyledLoader>
			<StyledBaseCircle />
			<StyledAnimatingCircle />
		</StyledLoader>
	</StyledLoaderHolder>
);
