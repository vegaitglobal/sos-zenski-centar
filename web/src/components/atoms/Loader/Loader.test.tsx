import renderer from 'react-test-renderer';
import Loader from './index';
import { renderWithTheme, componentWithTheme } from '../../../utils/testsWithTheme';

describe('Loader', () => {
	it('Renders Loader', () => {
		const { getByTestId } = renderWithTheme(<Loader />);

		const element = getByTestId('Loader');

		expect(element).toBeTruthy();
	});

	it('Renders correctly', () => {
		const tree = renderer.create(componentWithTheme(<Loader />)).toJSON();

		expect(tree).toMatchSnapshot();
	});
});
