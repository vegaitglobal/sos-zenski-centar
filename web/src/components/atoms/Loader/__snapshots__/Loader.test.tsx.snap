// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Loader Renders correctly 1`] = `
<div
  className="sc-bdnxRM jPJBih"
  data-testid="Loader"
>
  <svg
    className="sc-gtsrHT bPNifp"
    height={60}
    width={60}
  >
    <circle
      className="sc-dlnjwi sc-hKFxyN jPHmSv jesWCz"
      cx={30}
      cy={30}
      r={26}
    />
    <circle
      className="sc-dlnjwi sc-eCApnc jPHmSv dGQICf"
      cx={30}
      cy={30}
      r={26}
    />
  </svg>
</div>
`;
